// plexx/framework/src/nodes/transformation/translationNode.ts
/// <reference path="../../helper/references.ts" />

namespace Plexx {

    export interface TranslationInterface {
        translation?: number[];
    }

    export class TranslationNode extends TransformationNode {
        translation: number[];

        constructor(values: TranslationInterface) {
            super("TranslationNode");
            this.translation = values.translation || [0, 0];
        }

        public setTranslation(translation: number[]) {
            this.translation = translation;
        }

        public toString(): string {
            let translationString = this.getName() + " { ";
            translationString += "translation: " + this.translation + "; ";
            translationString += "}";
            return translationString;
        }

        public clone(): TranslationNode {
            return new TranslationNode({
                translation: this.translation
            });
        }

        public getTransformationMatrix(): Plexx.Mathlib.Mat3 {
            let translationMatrix: Plexx.Mathlib.Mat3 = new Plexx.Mathlib.Mat3([1, 0, 0, 0, 1, 0, this.translation[0], this.translation[1], 1]);
            return translationMatrix;

        }

        public getInverseTransformationMatrix(): Plexx.Mathlib.Mat3 {
               let translationMatrix: Plexx.Mathlib.Mat3 = new Plexx.Mathlib.Mat3([1, 0, 0, 0, 1, 0, -this.translation[0], -this.translation[1], 1]);
            return translationMatrix;
        }
    }
}
